freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaTopologyService.java

```diff
@@ -0,0 +1,163 @@
+package com.sequenceiq.freeipa.service.freeipa.flow;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.sequenceiq.freeipa.client.FreeIpaClient;
+import com.sequenceiq.freeipa.client.FreeIpaClientException;
+import com.sequenceiq.freeipa.client.model.TopologySegment;
+import com.sequenceiq.freeipa.client.model.TopologySuffix;
+import com.sequenceiq.freeipa.entity.InstanceMetaData;
+import com.sequenceiq.freeipa.entity.Stack;
+import com.sequenceiq.freeipa.service.freeipa.FreeIpaClientFactory;
+import com.sequenceiq.freeipa.service.stack.StackService;
+import org.springframework.stereotype.Service;
+
+import javax.inject.Inject;
+import java.util.Iterator;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@Service
+class FreeIpaTopologyService {
+
+    private static final Integer MAX_REPLICATION_CONNECTIONS = 4;
+
+    @Inject
+    private StackService stackService;
+
+    @Inject
+    private FreeIpaClientFactory freeIpaClientFactory;
+
+    public void updateReplicationTopology(Long stackId) throws FreeIpaClientException {
+        Stack stack = stackService.getByIdWithListsInTransaction(stackId);
+        Set<String> allNodesFqdn = stack.getNotDeletedInstanceMetaDataSet().stream()
+                .map(InstanceMetaData::getDiscoveryFQDN)
+                .collect(Collectors.toSet());
+        FreeIpaClient freeIpaClient = freeIpaClientFactory.getFreeIpaClientForStack(stack);
+        Set<TopologySegment> topology = generateTopology(allNodesFqdn).stream()
+                .map(pair -> {
+                    TopologySegment s = new TopologySegment();
+                    s.setDirection("both");
+                    s.setLeftNode(pair.getLeft());
+                    s.setRightNode(pair.getRight());
+                    s.setCn(String.format("%s-to-%s", pair.getLeft(), pair.getRight()));
+                    return s;
+                }).collect(Collectors.toSet());
+        List<TopologySuffix> topologySuffixes = freeIpaClient.findAllTopologySuffixes();
+        for (TopologySuffix topologySuffix : topologySuffixes) {
+            createMissingSegments(freeIpaClient, topologySuffix.getCn(), topology);
+            removeExtraSegements(freeIpaClient, topologySuffix.getCn(), topology);
+        }
+    }
+
+    @VisibleForTesting
```

Comment:
if it's public, do we need this annotation?

Reply:
Thanks for catching this, I meant to make it package private, so I made that correction.